File: backend/app/schemas/task.py
from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field


class Platform(str, Enum):
    BILIBILI = "bilibili"
    DOUYIN = "douyin"
    TWITTER = "twitter"
    YOUTUBE = "youtube"
    REDDIT = "reddit"


class DeliveryMode(str, Enum):
    DIRECT = "direct"
    DOWNLOAD = "download"


class TaskStatus(str, Enum):
    PENDING = "pending"
    PARSING = "parsing"
    DOWNLOADING = "downloading"
    MERGING = "merging"
    UPLOADING = "uploading"
    SUCCESS = "success"
    FAILED = "failed"


class ResultType(str, Enum):
    DIRECT = "direct"
    DOWNLOAD = "download"
    SPLIT_STREAMS = "split_streams"


class TaskResult(BaseModel):
    result_type: ResultType
    file_id: str | None = None
    file_name: str | None = None
    content_type: str | None = None
    download_url: str | None = None
    direct_url: str | None = None
    redirect_url: str | None = None
    proxy_url: str | None = None
    video_url: str | None = None
    video_redirect_url: str | None = None
    video_proxy_url: str | None = None
    audio_url: str | None = None
    audio_redirect_url: str | None = None
    audio_proxy_url: str | None = None
    placeholder: bool = False
    created_at: datetime
    file_size: int | None = None
    expires_note: str | None = None


class TaskRecord(BaseModel):
    task_id: str
    source_url: str
    platform: Platform
    delivery_mode: DeliveryMode = DeliveryMode.DIRECT
    status: TaskStatus
    progress: int = Field(default=0, ge=0, le=100)
    title: str
    message: str
    requires_merge: bool = False
    direct_playable: bool = False
    created_at: datetime
    updated_at: datetime
    result: TaskResult | None = None
    error_message: str | None = None
    uploader: str | None = None
    duration: int | None = None
    thumbnail: str | None = None
    extractor: str | None = None

File: frontend/src/types/task.ts
export type Platform = 'bilibili' | 'douyin' | 'twitter' | 'youtube' | 'reddit'
export type DeliveryMode = 'auto' | 'direct' | 'download'

export type TaskStatus =
  | 'pending'
  | 'parsing'
  | 'downloading'
  | 'merging'
  | 'uploading'
  | 'success'
  | 'failed'

export type ResultType = 'direct' | 'download' | 'split_streams'

export interface TaskResult {
  result_type: ResultType
  file_id: string | null
  file_name: string | null
  content_type: string | null
  play_url: string | null
  download_url: string | null
  direct_url: string | null
  redirect_url: string | null
  proxy_url: string | null
  video_url: string | null
  video_redirect_url: string | null
  video_proxy_url: string | null
  audio_url: string | null
  audio_redirect_url: string | null
  audio_proxy_url: string | null
  placeholder: boolean
  created_at: string
  file_size: number | null
  expires_note: string | null
}

export interface TaskRecord {
  task_id: string
  source_url: string
  platform: Platform
  delivery_mode: DeliveryMode
  status: TaskStatus
  progress: number
  title: string
  message: string
  requires_merge: boolean
  direct_playable: boolean
  created_at: string
  updated_at: string
  result: TaskResult | null
  error_message: string | null
  uploader: string | null
  duration: number | null
  thumbnail: string | null
  extractor: string | null
}

export interface ParseRequest {
  url: string
  delivery_mode?: DeliveryMode
}

export interface ParseAcceptedResponse {
  task: TaskRecord
  note: string
}

export interface HealthResponse {
  status: string
  app_name: string
  cleanup_interval_hours: number
  cleanup_retention_hours: number
  api_public_origin: string
  yt_dlp_available: boolean
  ffmpeg_available: boolean
  default_delivery_mode: DeliveryMode
  supported_platforms: string[]
}
